WAF: Move module initialization to the WAF package

nateweller opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Moves the waf module initialization into the WAF package, so that it is run by any plugin that uses `$config->ensure( 'waf' )`.

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

1201069996155224-as-1203517876254355

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* TBD

nateweller committed:
Move the initialization of the WAF to the package

nateweller committed:
changelog

nateweller committed:
WAF: Include all vendor files in the mirror repo build (#27850)

dkmyta commented on projects/packages/waf/src/class-waf-initializer.php:
```diff
@@ -17,11 +17,25 @@ class Waf_Initializer {
+			return;
+		}
+
+		// Triggers when the Jetpack plugin is updated
+		add_action( 'upgrader_process_complete', array( Waf_Runner::class, 'update_waf' ) );
```
This appears specific to Jetpack, do we know if this will work here as intended, or if maybe we'll need to add this line to each plugin that actually installs the WAF so the trigger actually takes effect on plugin update?

jwidavid replied:
That hook should work regardless of the plugin, although the comment could probably be updated.

nateweller replied:
Looking into the action details, this runs any time anything in WordPress updates. If we only want to run this when one of our plugins are updated, I can add a wrapper function that only runs the `update_waf` method if we just updated the plugin that is currently using the package.

nateweller replied:
Like so: 04f6fac46c8e

dkmyta reviewed (commented):
This appears to work well for Protect and things continue to work well on the Jetpack side.

My testing process included the following:
- Activated Protect (no plan) enabled the Firewall
- Verified the request to `?test=<script>` is not blocked
- Upgraded without issue, toggled Firewall off/on, and verified that the request is now blocked
- Deactivated Protect, activated Jetpack and performed the same steps with success
- Enabled Protect in addition to Jetpack and ensured that there are no conflicts and that updates and functionality remained in parity
- Throughout each step doubled checked that all actions remained functional
- Tested using Jurassic Tube to cover all changes
- Tested using Jurassic Ninja and Pressable w/ the Jetpack Beta tester and this branch to cover all things Protect related

Is there anything else critical to consider testing that wouldn't have been covered in the above steps with regard to plugin (Jetpack or Protect) or package functionality or the various processes involved?

One other thought came to mind, will this all be backward compatible? What if, for example, someone has the latest version of Protect installed and active, but a version of Jetpack active from prior to the change - could this potentially cause any issues with WAF functionality, or in-plugin settings changes? Testing didn't seem to indicate any issues, but figured it's worth ensuring we have it covered.

jwidavid commented on projects/plugins/protect/.gitattributes:
```diff
@@ -13,6 +13,7 @@ vendor/autoload_packages.php                    production-include
 vendor/automattic/**                            production-include
 vendor/composer/**                              production-include
 vendor/jetpack-autoloader/**                    production-include
+vendor/wikimedia/aho-corasick/**                production-include
```
Why is this being added?

dkmyta replied:
The discussion can be found here: p1670548960264089-slack-CBG1CP4EN

TL;DR as I understood it, that package dependency wasn't being added to Protect's list of dependencies and it was causing a fatal somewhere down the line.

nateweller committed:
Only update waf on relevant plugin upgrades

nateweller committed:
Update hook comment for better accuracy

nateweller commented:
@dkmyta @jwidavid I've addressed all comments, if you could peruse the one additional change to the `upgrader_process_complete` hook :+1:

Honestly best way I've found to test it is `error_log`s to ensure it is executing at the right time 😅

nateweller committed:
Fix conditional

nateweller committed:
Fix indentation

nateweller committed:
Improve plugins check conditional

dkmyta reviewed (approved):
Works like a charm!

sdixon194 committed:
Merge remote-tracking branch 'origin/trunk' into fix/protect-waf-init

sdixon194 reviewed (approved):
Looks like the module still loads from Jetpack's end 👍
